Tasks/Common/artifacts-common/serviceConnectionUtils.ts

```diff
@@ -0,0 +1,128 @@
+import * as tl from "azure-pipelines-task-lib/task";
+
+export interface IExternalPackageSource {
+    /**
+     * The protocol-specific URI used to access the package source
+     */
+    uri: string;
+}
+
+export abstract class ServiceConnection
+{
+    constructor(
+        public packageSource: IExternalPackageSource,
+        public authType: ServiceConnectionAuthType) {
+    }
+}
+
+export class TokenServiceConnection extends ServiceConnection
+{
+    constructor(
+        public packageSource: IExternalPackageSource,
+        public token: string)
+    {
+        super(packageSource, ServiceConnectionAuthType.Token);
+    }
+}
+
+export class UsernamePasswordServiceConnection extends ServiceConnection
+{
+    constructor(
+        public packageSource: IExternalPackageSource,
+        public username: string,
+        public password: string)
+    {
+        super(packageSource, ServiceConnectionAuthType.UsernamePassword);
+    }
+}
+
+export class ApiKeyServiceConnection extends ServiceConnection
+{
+    constructor(
+        public packageSource: IExternalPackageSource,
+        public apiKey: string)
+    {
+        super(packageSource, ServiceConnectionAuthType.ApiKey);
+    }
+}
+
+export enum ServiceConnectionAuthType
+{
+    Token, 
+    UsernamePassword,
+    ApiKey
+}
+
+/**
+ * Parses service connections / service endpoints from a task input into strongly typed objects containing the URI and credentials.
+ * 
+ * @param endpointsInputName The name of the task input containing the service connections endpoints
+ */
+export function getPackagingServiceConnections(endpointsInputName: string): ServiceConnection[]
+{
+    let endpointNames = tl.getDelimitedInput(endpointsInputName, ',');
+
+    if (!endpointNames || endpointNames.length === 0)
+    {
+        return [];
+    }
+
+    let serviceConnections: ServiceConnection[] = [];
+    endpointNames.forEach((endpointName: string) => {
+        let uri = tl.getEndpointUrl(endpointName, false);
+        let endpointAuth = tl.getEndpointAuthorization(endpointName, true);
```

Comment:
For some service connections we also have a data parameter, for example Pip and Twine (and now also Maven) we get the endpointname data parameter which we use while generating the authorizations in the respective files.

These parameters come in the form ENDPOINT_DATA_\<endpointName\>_\<key\>. Should we add something like an additional properties field to the service connection types and populate them with all the data keys we find? If we want other tasks to also use this library we would have to add some way to enable this functionality or else they will have to write custom logic to just get the data parameters. Let me know your thoughts

Reply:
Seems fine, feel free to do this as needed.